src/databricks/labs/ucx/azure/access.py

```diff
@@ -54,6 +56,43 @@ def __init__(
             "Storage Blob Data Owner": Privilege.WRITE_FILES,
             "Storage Blob Data Reader": Privilege.READ_FILES,
         }
+        self._permission_levels = {
+            "Microsoft.Storage/storageAccounts/blobServices/containers/write": Privilege.WRITE_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write": Privilege.WRITE_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/read": Privilege.READ_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read": Privilege.READ_FILES,
+        }
+
+    def _get_permission_level(self, permission_to_match: str) -> Privilege | None:
+        for each_level, privilege_level in self._permission_levels.items():
+            # Check for storage blob permission with regex to account for star pattern
+            match = re.search(permission_to_match, each_level)
+            # If a write permission is found, no need to check for read permissions
+            if match:
+                return privilege_level
+        return None
+
+    def _get_custom_role_privilege(self, role_permissions: list[str]) -> Privilege | None:
+        # If both read and write privileges are found, only write privilege will be considered
+        higher_privilege = None
+        for each_permission in role_permissions:
+            if each_permission.startswith('Microsoft.Storage'):
+                privileges = self._get_permission_level(each_permission)
+                if privileges and privileges == Privilege.READ_FILES:
+                    higher_privilege = Privilege.READ_FILES
+                elif privileges and privileges == Privilege.WRITE_FILES:
+                    higher_privilege = Privilege.WRITE_FILES
+                    break
+        return higher_privilege
+
+    def _get_role_privilege(self, role_assignment: AzureRoleAssignment) -> Privilege | None:
+        privilege = None
+        # Check for custom role permissions on the storage accounts
+        if role_assignment.role_permissions:
+            privilege = self._get_custom_role_privilege(role_assignment.role_permissions)
```

Comment:
Hi @JCZuurmond Thanks for reviewing the code. I am working on the comments, once done I will create a new PR and share the details with you. 
For this one, I have a question. How the existing functionality for built-in roles work is, if you have 2 role assignments (Blob Reader built-in role and Blob Contributor built-in role) to a service principal (sp1), then the output will be 2 rows -- one row for sp1 read files, one role for sp1 write files.
Similarly, with the new changes, if I have 2 role assignments -- one Blob Contributor built-in role, one read access through custom role, then the output will be 2 rows -- one row for sp1 write files, one role for sp1 read files.
Please let me know if I should handle both the scenarios (existing and new) to reflect only the higher privilege.

Reply:
Ah, I now see that the existing code indeed creates multiple `StoragePermissionMapping` for the same principal when it is assigned multiple roles. So, let's keep that the same for your code changes.

However, we need to figure out if this goes well. Reading the code in `ServicePrincipalMigration` I don't see deduplication of the storage permissions and the create storage credentials are named after service principals. This creates a collision, with a chance of creating the credential with a lower permission (read instead of write). 

@nkvuong : Could you chime here? Do I read that correctly?

Reply:
that looks correct to me. the risk of collision was lower before as it is less likely for a principal to have both contributor & read access (it is redundant). if we now include custom roles, the collision risk is higher, so I would prefer if the changes are added in this PR

Reply:
@JCZuurmond @nkvuong - I have made changes to handle this scenario in the last commit. Please review.